File: 19_数学/中位数众数/229. 求众数 II.ts
// 给定一个大小为 n 的整数数组，找出其中所有出现超过 ⌊ n/3 ⌋ 次的元素。
function majorityElement(nums: number[]): number[] {
  const count = (nums: number[], target: number) => {
    let res = 0
    const n = nums.length
    for (let i = 0; i < n; i++) {
      nums[i] === target && res++
    }
    return res
  }
  const n = nums.length
  if (n < 1) return []
  if (n < 2) return nums

  let count1 = 0
  let count2 = 0
  let candidate1 = 0
  let candidate2 = 1
  for (const num of nums) {
    if (num === candidate1) count1++
    else if (num === candidate2) count2++
    else if (count1 === 0) (candidate1 = num), (count1 = 1)
    else if (count2 === 0) (candidate2 = num), (count2 = 1)
    else count1--, count2--
  }

  const res: number[] = []
  count(nums, candidate1) > n / 3 && res.push(candidate1)
  count(nums, candidate2) > n / 3 && res.push(candidate2)
  return res
}

console.log(majorityElement([1, 1, 1, 3, 3, 2, 2, 2]))
export {}

File: 19_数学/众数/229. 求众数 II.py
# !给定一个大小为 n 的整数数组，找出其中所有出现超过 ⌊ n/3 ⌋ 次的元素。
# 尝试设计时间复杂度为 O(n)、空间复杂度为 O(1)的算法解决此问题
from collections import Counter
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        """counter肯定是最直接的
        
        O(n)
        """
        counter = Counter(nums)
        res = []
        for num, count in counter.items():
            if count > len(nums) // 3:
                res.append(num)
        return res

    def majorityElement2(self, nums: List[int]) -> List[int]:
        """摩尔投票法"""
        res = []
        count1, count2 = 0, 0
        cand1, cand2 = int(1e20), int(1e30)
        for num in nums:
            if num == cand1:
                count1 += 1
            elif num == cand2:
                count2 += 1
            elif count1 == 0:
                cand1 = num
                count1 = 1
            elif count2 == 0:
                cand2 = num
                count2 = 1
            else:
                count1 -= 1
                count2 -= 1
        if nums.count(cand1) > len(nums) // 3:
            res.append(cand1)
        if nums.count(cand2) > len(nums) // 3:
            res.append(cand2)
        return res
